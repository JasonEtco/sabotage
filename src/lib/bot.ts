import { Client, ClientOptions, Message } from 'discord.js'
import { Signale } from 'signale'
import MessageError from './message-error'

export class Bot extends Client {
  public log: Signale

  constructor(opts?: ClientOptions) {
    super(opts)
    this.log = new Signale()
  }

  /**
   * Returns true if the message was sent by the bot
   */
  public isSelf(msg: Message) {
    return msg.author.id === this.user!.id
  }

  /**
   * Load a hash of bot functions
   */
  public async load(funcs: { [key: string]: Function }) {
    // Register all the bot functions
    for (const func in funcs) {
      this.log.info(`Registering function [${func}]`)
      await funcs[func](this)
    }
  }

  /**
   * Respond with a formatted error message if the handler throws
   * an instance of MessageError
   */
  public async respondWithError(err: any, msg: Message) {
    this.log.error(err)

    if (!(err instanceof MessageError)) return

    return msg.channel.send('Error:', {
      embed: {
        color: 0xff0000,
        description: err.message,
      },
    })
  }

  /**
   * An extended `client.on`, this allows for handlers to declare
   * a RegEx to match before triggering the handler and automatically
   * ignore messages sent by the bot
   */
  public onMatch(
    reg: RegExp,
    handler: (msg: Message, match: RegExpMatchArray) => Promise<unknown>
  ) {
    return this.on('message', async (msg) => {
      // Ignore message sent by the bot
      if (this.isSelf(msg)) return

      // Ensure that we've found a match
      const str = msg.cleanContent.toString()
      const match = str.match(reg)
      if (!match) return

      return handler(msg, match).catch((err: any) =>
        this.respondWithError(err, msg)
      )
    })
  }

  /**
   * Act on "command" messages starting with a `.`
   * chracter, like .deploy <args>.
   */
  public command(
    command: string,
    handler: (args: string, msg: Message) => Promise<unknown>
  ) {
    const COMMAND_KEY = '\\.'
    const reg = new RegExp(`^${COMMAND_KEY}${command}(?:$|\\s(.*))`)
    return this.onMatch(reg, async (msg, match) => {
      const args = match[1] || ''
      return handler(args, msg)
    })
  }
}
