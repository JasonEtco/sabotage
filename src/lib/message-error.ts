/**
 * A custom error class that is intended to be sent to Discord
 */
export default class MessageError extends Error {}
