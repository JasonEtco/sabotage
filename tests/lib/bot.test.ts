import { Signale } from 'signale'
import { Message, ClientUser, User } from 'discord.js'
import { Bot } from '../../src/lib/bot'

describe('Bot', () => {
  let bot: Bot

  beforeEach(() => {
    bot = new Bot()
    bot.log = new Signale({ disabled: true })
    bot.user = { id: 'BOT' } as ClientUser
  })

  describe('#isSelf', () => {
    it('returns `true` if the message was sent by the bot', () => {
      expect(bot.isSelf({ author: { id: bot.user!.id } } as Message)).toBe(true)
    })

    it('returns `false` if the message was sent not by the bot', () => {
      expect(
        bot.isSelf({ author: { id: 'JASON HELLO' } as User } as Message)
      ).toBe(false)
    })
  })

  describe('#load', () => {
    it('loads the provided functions', async () => {
      bot.log.info = jest.fn()
      const funcs = {
        someFunction: jest.fn(async (bot) => bot.log.info('I loaded!')),
      }
      await bot.load(funcs)
      expect(funcs.someFunction).toHaveBeenCalled()
      expect(bot.log.info).toHaveBeenCalled()
    })
  })

  describe('#replyWithError', () => {
    it('calls msg.reply with the error message', async () => {
      const msg = {
        reply: jest.fn() as unknown,
      } as Message
      await bot.replyWithError('Oh no!', msg)

      expect(msg.reply).toHaveBeenCalled()
      const call = (msg.reply as jest.Mock).mock.calls[0]
      expect(call).toMatchSnapshot()
    })
  })

  describe('#onMatch', () => {
    it('registers a `message` event handler that ignores messages sent by itself', () => {
      const spy = jest.fn(async () => {})
      bot.onMatch(/.*/, spy)
      bot.emit('message', { author: { id: bot.user!.id } } as Message)
      expect(spy).not.toHaveBeenCalled()
    })

    it('registers a `message` event handler with a RegEx', () => {
      const spy = jest.fn(async () => {})
      bot.onMatch(/^test$/, spy)
      // Should call the spy
      bot.emit('message', {
        cleanContent: 'test',
        author: { id: 'JASON' },
      } as Message)
      // Should not call the spy
      bot.emit('message', {
        cleanContent: 'nope',
        author: { id: 'JASON' },
      } as Message)
      expect(spy).toHaveBeenCalledTimes(1)
    })
  })

  describe('#command', () => {
    it('runs the handler for a matching command message', () => {
      const spy = jest.fn()
      bot.command('hello', spy)
      const msg = {
        cleanContent: '.hello Jason',
        author: { id: 'JASON' },
      } as Message
      bot.emit('message', msg)
      expect(spy).toHaveBeenCalled()
      expect(spy).toHaveBeenCalledWith('Jason', msg)
    })

    it('runs the handler for a matching command message without arguments', () => {
      const spy = jest.fn()
      bot.command('hello', spy)
      const msg = { cleanContent: '.hello', author: { id: 'JASON' } } as Message
      bot.emit('message', msg)
      expect(spy).toHaveBeenCalled()
      expect(spy).toHaveBeenCalledWith('', msg)
    })

    it('does not run the handler for a partially matching command', () => {
      const spy = jest.fn()
      bot.command('hello', spy)
      const msg = {
        cleanContent: '.hellofresh',
        author: { id: 'JASON' },
      } as Message
      bot.emit('message', msg)
      expect(spy).not.toHaveBeenCalled()
    })
  })
})
