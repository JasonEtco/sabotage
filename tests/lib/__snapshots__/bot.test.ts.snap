// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bot #replyWithError calls msg.reply with the error message 1`] = `
Array [
  Object {
    "embed": Object {
      "color": 16711680,
      "description": "Oh no!",
    },
  },
]
`;
